### ISSUE TITLE
Diagnostic for unused call specification

### ISSUE THREAD

**Issue by thomaslevesque** [FakeItEasy/FakeItEasy#490]
The analyzer should warn about unused call specifications.

A call to `A.CallTo(...)` (or to any other method of the fluent API used to specify the call) should always be followed by another call, or the result should be assigned to a variable. Otherwise, it means that the call specification is not used, which is almost certainly a mistake.

The diagnostic following shown for the following calls:

```
IAnyCallConfigurationWithNoReturnTypeSpecified.WithReturnType<TMember>
IArgumentValidationConfiguration<TInterface>.WhenArgumentsMatch
ArgumentValidationConfigurationExtensions.WithAnyArguments<TInterface>
IWhereConfiguration<T>.Where
WhereConfigurationExtensions.Where
A.CallTo // (all 3 overloads)
Fake<T>.CallsTo // both overloads
Fake<T>.AnyCall
```

Related to #489 

---

TODO:

- [ ] replace attributes with member to diagnostic map

**Comment by drauch** [comment]
I would also like R# analyzer support for this. I know #505 has been closed, however, R# 10 now features a better attribute describing exactly the situation at hand:

https://www.jetbrains.com/help/resharper/10.0/Reference__Code_Annotation_Attributes.html?origin=old_help#MustUseReturnValueAttribute

Please add `MustUseReturnValueAttribute` to the `A.CallTo()` method!

**Comment by drauch** [comment]
I've created a pull request for R# in https://github.com/FakeItEasy/FakeItEasy/pull/629

**Comment by blairconrad** [comment]
@drauch, thanks for your interest. Your point about moving to another issue is a good one, but because I think it'll be of interest, let me ask this here:

What benefit do the ReSharper annotations provide over the code analyzer? (I know, annotations can be used with older VS, but only if have ReSharper 10. Aside from that.)

If the code analyzer were available right now, what impetus would there be to add the annotations?

**Comment by drauch** [comment]
Of course the main advantage is the immediate availability. Nobody really knows when the Roslyn Analyzer is going to be "done done & released".

Other than that, it is nice to have R# InspectCode available on build agents where MSBuild 14 is not yet installed (this is of personal interest for us). In the long run, we're fine with a released version of the analyzer as well, however, I'm smelling there will be no analyzer for FakeItEasy 1.x and our pull request #629  is immediately mergeable (and is easily removed later on as well). So... :-)

**Comment by thomaslevesque** [comment]
> Nobody really knows when the Roslyn Analyzer is going to be "done done & released".

It already exists as a POC, there's little more to do to call it done.

Anyway, I agree that having both a Roslyn analyzer and the R# annotation is useful. Not everyone has R#, and not everyone has VS2015, so providing support for both makes it available to more people. The only drawback is that if you have both R# and the analyzer enabled, the warning will appear twice:

![screenshot](http://i.imgur.com/cAvDT1x.png)

BTW, could you change the message in the attribute to make it match [the one from the analyzer](https://github.com/thomaslevesque/FakeItEasy/blob/analyzer/src/FakeItEasy.Analyzers/FakeItEasy.Analyzers/Resources.resx#L125)?

**Comment by drauch** [comment]
@thomaslevesque: PR updated, couldn't change the message to match the Roslyn analyzer's message exactly, as R# doesn't support to insert parameters into the message (afaik).

**Comment by blairconrad** [comment]
@drauch said 

>  however, I'm smelling there will be no analyzer for FakeItEasy 1.x

Nor would I expect a ReSharper annotation for FakeItEasy 1.x. We long ago closed that line to all but the most severe of bugs. Any changes will be considered for the 2.x line.

**Comment by blairconrad** [comment]
@thomaslevesque said

> The only drawback is that if you have both R# and the analyzer enabled, the warning will appear twice

I'd say that there's an additional (admittedly small) burden. If we have two ways to provide warnings of this this type, then we have to maintain two ways to provide warnings of this type.
I expect that means both ensuring that the two mechanisms continue to work and maintaining feature parity: every time we change the analyzer, we'd at least have to consider making a parallel change to the annotations.

**Comment by adamralph** [comment]
I'm still of the opinion that we should focus our efforts on Roslyn analyzers for providing static analysis. They are promising to be the ubiquitous solution for this in the .NET space and are likely to ultimately have wide support across the (growing) .NET IDE/posh text editor ecosystem.

> Of course the main advantage is the immediate availability. Nobody really knows when the Roslyn Analyzer is going to be "done done & released".

**Comment by drauch** [comment]
> That is true, but all the more reason for focusing our (very limited) time on that.
> All good reasons for us focusing on the analyzer as our static analysis solution.

You realize, that you do not have to waste any time on the R# static analysis, as it is basically mergeable right now? Moreover, my PR is easily undone if at a later point your predictions about Roslyn analyzers are becoming reality.

Hope you reconsider your direction, please keep me posted.

**Comment by blairconrad** [comment]
@drauch, it may seem like we don't have to expend time on the Resharper static analysis, since the PR is mergeable now, but I think there may be an aspect to it that you're not considering. Often, features require care and feeding after the initial implementation. In the case of the Resharper static analysis, in order to support it properly, we'd have to make sure it keeps being relevant as the code changes, and doesn't conflict with other analyzers. Probably this wouldn't be a big deal, but it's not quite nothing.
And I'd say that by accepting the PR, we're opening the door to Resharper attributes. I'd expect more requests for changes - for example, `A.Dummy`'s return value could be `NotNull`, and so on. Each of these changes would probably be simple, but it's a bit of a commitment. And the effort expended may not be worth it just to support a _secondary_ way to warn people about FakeItEasy misuse.

Finally, removing the attributes would not be so easy in the future, as I'd considered that to be a breaking change—we would've been providing a service and then all of a sudden not. As such, we'd be loathe to make the change, and would need to wait for (or force) a major release.

I'm not saying that these are sufficient reasons for not taking the PR, but they do provide food for thought, and explain my hesitation (and I suspect maybe some of @adamralph's, but I shouldn't put words in his mouth). I'm sorry if this procedure is frustrating. It may seem like we don't appreciate your interest, but it's not so. Sometimes it just takes a while to find what we think is the best path forward.

**Comment by adamralph** [comment]
Thanks @blairconrad, that echoes my thoughts precisely. /cc @drauch

**Comment by thomaslevesque** [comment]
> for example, `A.Dummy`'s return value could be `NotNull`

That's actually a very interesting example. Nullability checking is something that R# does pretty well and that's easy to configure with annotations, but it's not something our analyzer would do, because it's:
1. out of scope (not specific to FakeItEasy)
2. _very_ hard

> Finally, removing the attributes would not be so easy in the future, as I'd considered that to be a breaking change—we would've been providing a service and then all of a sudden not. As such, we'd be loathe to make the change, and would need to wait for (or force) a major release.

I'm not sure I'd call that a breaking change... It wouldn't change the behavior of the library, so it wouldn't break anyone's code.

Regarding the effort of maintaining the annotations, I'm not very worried about that, and I don't mind taking care of it myself if you guys are afraid it will be a burden.

So, personally I'm still in favor of including R# annotations, but I won't fight over this.

**Comment by adamralph** [comment]
IMHO the [ReSharper Gallery](https://resharper-plugins.jetbrains.com/) is the solution for R# annotations.

It already has a [JetBrains.ExternalAnnotations](https://resharper-plugins.jetbrains.com/packages/JetBrains.ExternalAnnotations/) package which includes annotations for many popular libraries. I'm not sure if it's OSS, but if so, then annotations for FIE could be PR'd to it. If not, then whoever is interested, can create a FakeItEasy annotations package and upload it there. 

@drauch any interest? You can create your own external annotations package there and go wild with your attributes! :wink:

**Comment by adamralph** [comment]
I'd also be happy for us to include a list of third party FIE packages in our docs, and that could list the relevant R# gallery package(s).

**Comment by blairconrad** [comment]
@drauch, I'd be happy to contribute to the external annotations.

**Comment by drauch** [comment]
A colleague of mine proposed to open-source JetBrains.ExternalAnnotations and thereby enable pull requests. Fortunately, JetBrains reacted within a single day (https://youtrack.jetbrains.com/issue/RSRP-456547), and we are able to contribute now at https://github.com/JetBrains/ExternalAnnotations

Still, [I'm not a huge fan](https://i.imgur.com/wvQm5.jpg): due to the fact that those annotations can get out-of-sync with FakeItEasy pretty easily. The external annotation story is becoming feasible for users only if JetBrains decides to implement https://youtrack.jetbrains.com/issue/RSRP-427415. Until that point I'd still prefer the inline R# attributes (although that ship seems to have sailed yesterday...)

**Comment by blairconrad** [comment]
@drauch, congratulations on your success!

> [I'm not a huge fan](https://i.imgur.com/wvQm5.jpg)

Nice. If you're looking for an alternative sometime, [consider this](https://xkcd.com/1378/).

>  those annotations can get out-of-sync with FakeItEasy pretty easily.

The problem probably isn't as bad as you think. We are loathe to make breaking changes in the library, so the annotations may _lag_ FakeItEasy, if we add surface area to the public API, but that's probably most of what there is to worry about.

Regarding RSRP-427415, it seems there's hope there already, no? Or do I misunderstand [Slava Trenogin's comment](https://youtrack.jetbrains.com/issue/RSRP-427415#comment=27-1380453)?

**Comment by blairconrad** [comment]
@drauch, I did misunderstand your question and  Slava Trenogin's comment. I get it now.

**Comment by thomaslevesque** [comment]
Thanks for the feedback @adamralph!

**Comment by adamralph** [comment]
Taking this off the 2.0.0 milestone for now, since it doesn't *have* to be done for 2.0.0 to go live. If it's done, and we've resolved #666, by the time we're ready to release 2.0.0 then it can go back on the milestone.

**Comment by drauch** [comment]
FYI: the R# annotations already detected a bug in one of our test setup methods :-)

**Comment by blairconrad** [comment]
That's fantastic, @drauch!

**Comment by adamralph** [comment]
@drauch out of interest, what did the annotation catch? We'd love to capture as much as possible in our new Roslyn Analyzer so it would be great to find out where to focus our efforts.

**Comment by drauch** [comment]
A missing .MustHaveHappened() call in one of our specs has been detected.

**Comment by adamralph** [comment]
Aha, so our Roslyn analyzer what have caught it too :smiley:

**Comment by blairconrad** [comment]
@adamralph, indeed. The FakeItEasy annotations at JetBrains.ExternalAnnotations [that have been added so far](https://github.com/JetBrains/ExternalAnnotations/commits/master/Annotations/Misc/FakeItEasy) only have the annotation's I'd put in [my extension](https://github.com/blairconrad/FakeItEasy-annotations/), and are a subset of those the Roslyn analyzer had. (They don't cover `A.CallToSet`, for example.)

**Comment by adamralph** [comment]
I guess a measure of "having done our job" with the analyzer will be the redundancy of these things. :wink:

**Comment by blairconrad** [comment]
As far as I'm concerned, FakeItEasy.Annotations is already redundant. I just haven't told it. Will after 2.0 comes out. ALTHOUGH, there are 57 downloads. What will the community do without it?

**Comment by adamralph** [comment]
And it looks so shiny :sparkles: https://resharper-plugins.jetbrains.com/packages/FakeItEasy.Annotations :cry:

**Comment by blairconrad** [comment]
This issue has been fixed in [release 2.0.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/2.0.0).
